### ISSUE TITLE
[NEXT-1187] Link navigation with loading.tsx is not instant for dynamic pages

### ISSUE THREAD

**Issue by tonypizzicato** [vercel/next.js#43548]
### Verify canary release

- [X] I verified that the issue exists in the latest Next.js canary release

### Provide environment information

    Operating System:
      Platform: darwin
      Arch: arm64
      Version: Darwin Kernel Version 22.1.0: Sun Oct  9 20:14:30 PDT 2022; root:xnu-8792.41.9~2/RELEASE_ARM64_T8103
    Binaries:
      Node: 18.12.1
      npm: 8.19.2
      Yarn: 1.22.19
      pnpm: N/A
    Relevant packages:
      next: 13.0.6-canary.2
      eslint-config-next: 13.0.5
      react: 18.2.0
      react-dom: 18.2.0

### Which area of Next.js is affected? (leave empty if unsure)

App directory (appDir: true), Routing (next/router, next/navigation, next/link)

### Link to reproduction - Issues with a link to complete (but minimal) reproduction code will be addressed faster

https://github.com/tonypizzicato/next-13-loading

### To Reproduce

- pull the repo
- run `yarn install && yarn dev`
- open `localhost:3000` in the browser
- go to the "Careers" page
- reload the page
- set some throttling in devtools
- click different pages in the menu
- notice that "Careers" page loading state and route change are done instantly but with some delay for other pages

### Describe the Bug

When you have a dynamic page with `loading.tsx` the route change and showing the loading animation are instant only for the page, which was freshly loaded. For other dynamic pages it hits the server first, then shows the loading state
<img width="193" alt="Screenshot 2022-11-29 at 23 19 38" src="https://user-images.githubusercontent.com/640122/204660865-5d64b2bf-ef2f-4c4e-9d7b-5cf05944e3d1.png">

https://user-images.githubusercontent.com/640122/204661070-b9409f71-3814-45e6-bf56-923cdf4e2fc6.mov

### Expected Behavior

Instantly navigate to the desired dynamic page and show loading component

<sub>[NEXT-1187](https://linear.app/vercel/issue/NEXT-1187/link-navigation-with-loadingtsx-is-not-instant-for-dynamic-pages)</sub>

**Comment by Fredkiss3** [comment]
That is how it works, in order to do the loading state, your browser has to first make the request to nextjs server, then nextjs will first return the loading state, and after resolves your page. 

The thing is that in need, it is even slower that in production, because nextjs do a bigger amount of work. You could try to run it in production with `yarn build && yarn start`, but you'll find that it is not really instant, the slowness comes from the speed of your internet connection but also the distance between you and the server. If a normal request from you to the server is about 1s for example, then you'll see the loading state after 1s. But what is cool about the loading state if for example you make a request in your `page.tsx` which takes more that 5s (for example), you'll always see the loading state after 1s.

All of this is part of streaming rendering : https://beta.nextjs.org/docs/data-fetching/streaming-and-suspense#concepts

**Comment by tonypizzicato** [comment]
> at is how it works, in order to do the loading state, your browser has to first make the request to nextjs server, then nextjs will first return the loading state, and after resolves your page.

that totally makes sense, that's why i've added throttling for testing on the client and delay on the api side. but isn't it a job of page prefetch feature to get the loading state before navigating to it?
and if that's an intended behaviour, why does the first page loaded work differently (instant loading state on client every time)?
and what is the reason to use streaming api with loading.ts feature in next.js if it is only useful (after 1s in your example) for API endpoints that take some significant time? for the most applications with simple db access endpoints its seems useless and client loading state would be preferred.

More than that Next.js docs say that loading.tsx should be instant compared to custom Suspense boundaries
> Unlike loading.js, where navigation happens immediately and loading UI is displayed before the request to the server is complete. When manually defining Suspense Boundaries, navigation will happen after the new segment loads.
https://beta.nextjs.org/docs/routing/loading-ui#manually-defining-suspense-boundaries

**Comment by Fredkiss3** [comment]
> But what is cool about the loading state if for example you make a request in your page.tsx which takes more that 5s (for example), you'll always see the loading state after 1s.

As i said, for api endpoints that take some time to fetch (for example 5s) it would always take the same amount of time to show you the loading UI, the time it would take you to show the loading state is the time your browser to connect to the server. So if your browser takes 1s to connect to the server, the loading ui would show after 1s and not after 5s, so instead of waiting 5s to show your page, you get a loading state sooner. This is better for when you reload the page for ex.

And you're right, with prefetching it should be instant, since it has already the data... but it seems that when you have a loading state it prefetches twice ? 🤔
In the example below it does two requests to prefetch data (see by the `next-router-prefetch` header). I don't know if this is a bug or not : 

https://user-images.githubusercontent.com/38298743/204781992-7c49730d-5796-43ed-8c84-39b343b73877.mov


But prefetching isn't also a silver bullet. Take the case when your page hasn't finished prefetching the page, it would still take 1s before showing you the loading state and i think even if your page where static it would take the same amount of time : the time it takes for your browser to connect to your server and your server to serve you the static page.

Showing a loading state on the client would definitely make it faster,  but only if you don't issue a navigation : like you fetch on the client without navigating to another page i think... 🤔

**Comment by dumbravaandrei22** [comment]
Hmm, weird. For me, it worked on 13.0.6-canary.3 but it seems that I encountered the same issue on 13.1.1

**Comment by dumbravaandrei22** [comment]
Still happening in 13.1.3-canary.0

**Comment by dumbravaandrei22** [comment]
I think that the loading.tsx should be instant without any network calls. 
I know that the loading.tsx will be under a page request, but like @tonypizzicato said the website should prefetch the Links that are in the page.
That's the idea, from a user's POV, the user should receive instant feedback from the app. 
If the loading.tsx will take sometime to load (even a small one) it will be useless.

**Comment by eug-vs** [comment]
Absolutely agree with all problems raised in this issue, this is the only thing stopping our project from migrating to App Router - navigation creates just horrible experience with sometimes up to 300ms wait time before even the loading streams in (I know this can be solved by moving the server closer e.g. on edge, but that's a different story). Other side of the problem is that there's no way to display any other loading state on the UI while loading hasn't streamed in yet - you are forced into this UX.

I truly don't understand the reason behind this: if `loading.tsx` doesn't accept any props and is essentially static (usually just a bunch of skeletons), why can't we include it in client-side bundle (or at least opt-in to that)? Or event better - do not include it into a bundle, but fetch it *and cache properly on the client* - so that subsequent visits of the same page immediately show the fetch UI from client-side cache. That is probably how prefetch should work, but currently my experience is that it's working in a very random fashion (even in NextJS playground: https://app-dir.vercel.app/streaming/edge/product/2)

With `Suspense` it's a bit different because it allows for granular control, but even then the initial state where `Suspense` is showing fallback could be used with the same approach.

**Comment by devdreamsolution** [comment]
@tonypizzicato , I've resolved same issue with group route.
https://beta.nextjs.org/docs/routing/defining-routes#route-groups
Look at my app route structure.
![image](https://user-images.githubusercontent.com/54434088/230150290-0c42d957-6133-4ec9-be13-a8e5bc5fdf5f.png)
Dynamic page uses the nearest parent loading.tsx, but announcements must has another loading.tsx.
To resolve this problem, I used group route.
I hope this solution works well for you.

**Comment by mhesham32** [comment]
for me when I added a root loading.tsx file it was picked immediately after I started the navigation to another page while the loading.tsx file inside the other page's folder wasn't picked and I was getting the same behavior as @tonypizzicato before adding the root loading.tsx file.
also the other SSR page is a search page and it gets the search query from URL search params I was expecting the loading.tsx to appear when I change the params within the page to fetch new data but instead, it was displaying the old results after clicking on search and then suddenly the new results replace them which is not good for UX it was like the site was unresponsive

**Comment by mhesham32** [comment]
it's worth mentioning the above behavior was happening with next@13.2.4 but when I upgraded to 13.3 it was no longer happening

**Comment by devdreamsolution** [comment]
@mhesham32 , did you use client component in 13.3?
that has error in my end.
https://github.com/vercel/next.js/issues/43930#issuecomment-1501767146

**Comment by mhesham32** [comment]
Yes the navigation is triggered from a client component to a server page

**Comment by sourcecodes2** [comment]
We are also experiencing this issue, and it is still present on v13.3.

The concept of "loading" for the end user means anything that is delaying their request, whether it's due to server processing or network latency.

There should be no distinction to the end user on technical merit. This is purely a UX issue.

End users need immediate feedback, whether that's the data they've requested or feedback to say their request is being handled.

Currently, this implementation fails to do this.

This is a core part of the age-old business arguement for retaining impatient users that we work ever so hard to get, that cannot be overlooked.

The fact that the on-request loading feedback needs *loading on-request* makes this implementation feel like one step forward and two steps back.

We are very grateful for NextJS and we do appreciate that the AppDir is still in beta, but please make sure this UX issue is fixed before release.

**Comment by Fredkiss3** [comment]
It seems that to make loading instant you have to define it on the parent folder of the dynamic route like they do in [App dir playground repository](https://github.com/vercel/app-playground/tree/main/app/loading).


<img width="292" alt="image" src="https://user-images.githubusercontent.com/38298743/232018380-d825b7c0-93c8-4bac-9d1e-1b524ef718bd.png">

You can also use route groups like in @devdreamsolution comment : https://github.com/vercel/next.js/issues/43548#issuecomment-1497829622

**Comment by unleashit** [comment]
Glad to have found this issue for a partial fix. Hope the behavior changes for manual Suspense so that the fallback is shipped to the client and run at the start of the transition. That's the UX users are going to expect. I'm not sure what the behavior was pre app directory with getServersideProps. But in other isomorphic React projects I've worked on you could always show a loading state on the component level (not saying you couldn't in Next.js in a useEffect or getInitialProps, but my understanding is with GSSP it could only be done on the page level).

**Comment by Fredkiss3** [comment]
@unleashit my understanding is that pre-app router the loading fallback could be controlled by the user with router events (https://nextjs.org/docs/api-reference/next/router#routerevents) there is even an example used for showing a loading indicator : https://github.com/vercel/next.js/tree/canary/examples/with-loading

**Comment by unleashit** [comment]
Hey thanks for the reply but I'm pretty sure those router events were only for the full page (and not at the component level) as I was mentioning. I'm pretty sure Suspense fallbacks are intended as the solution, but I don't think it's implemented as well as it could be. It definitely an improvement over those router events, because you can control it on the "route segment" level with a child loading.tsx. That's better than it was, but better still would be to support the same behavior (fallback shown immediately on a client transition) in a manual Suspense fallback.

**Comment by Fredkiss3** [comment]
@unleashit those events where for page changes (or URL changes I think ?), it is in a useEffect so you can use it even at component levels, it is basically the same as `loading.tsx`.

I would not say that Suspense is implemented badly,but what it seems is that the client router is not aware of the `loading.tsx` components, or more likely, it wait for things to suspend (with `async/await` or `use`) to show the loading fallback, you can try for example to export a simple non async component and put a `loading.tsx` beside the page, it would not show the loading fallback (at least that's my understanding, I may be wrong).

I also recently inspected the RSC payload on navigation, and have seem to find that a `layout.tsx` ships with a loading indicator, so it may be the fact that if a layout for a page is located in the same level as a `loading.tsx` file, when the layout will load, next navigations will show the loading fallback instantly as it has already prefetched it.

I will try this again and post an update with my findings later.

**Comment by unleashit** [comment]
Yeah I agree with most of that and was only using the router transition as a reference for how things were UX wise with the NextJS experience. Suspense is an upgrade over that because it has more granular support for loading content from the server, vs only the top/page level for getServerSideProps.

But I think there definitely are some issues that people will agree on and will hopefully get ironed out. As someone mentioned above, as a "user", I would expect either the content or a loading transition to appear immediately after clicking a link, button, etc. Not after a delay, whether the delay is caused by the network or a data request somewhere, because the distinction doesn't really matter to them.

The experience is different with a loading.js and a manual Suspense. In the docs, the former is basically an automatic suspense for a page, but the loading behavior isn't the same. As an example, you can go to: https://jasongallagher.org/portfolio/tatteasy and click next/prev buttons. You should see a loader (a nested loading.tsx) just between the header and footer (* at least in Firefox, see below).

But then the behavior isn't the same with manual <Suspense />. I have one on the home page part way down the page. You can test by clicking on "work" in navbar. If you refresh with throttling on in dev tools, notice that the checkerboard of images never shows a spinner. The site is statically exported, but I've also tried this in SSR mode (and these are server components that query data) and the behavior is the same: almost never a spinner, even with throttling. In some of my testing while running in SSR mode, I did sometimes see a blip of a fallback in some places, but not in the way I would expect. IMO for a good UX, the fallback state should last for the full time that the HTML markup can't be shown in the browser regardless of why. 

I assume a static/SSG site doesn't show a suspense fallback because the RSC output is statically available. Yet the spinner works (albeit inconsistently) in Next's implementation of loading.js. It would be much nicer if the fallback was displayed consistently no matter the environment.

* I just noticed while writing this up that the loading.js behavior works better in Firefox (which I was mostly testing on) than Chrome. In Chrome, the spinner sometimes appears and sometimes doesn't. Although now I'm wondering if there was a change after upgrading to 13.3.1 because it seemed to always work before.

**Comment by Fredkiss3** [comment]
@unleashit , just an unrelated question though, how did you make your prefetch & link navigation link fetch `.txt` instead of the route ? 

I supposed you manually created a link component and used a middleware to rewrite those URLs to the actual route ? this is to fix cloudfare caching right ? 🤔 

![image](https://user-images.githubusercontent.com/38298743/233813241-c3387442-13e3-44d8-9a29-9167ee3b8305.png)

**Comment by Fredkiss3** [comment]
@dumbravaandrei22 maybe try to put a dummy Layout at `(feed)/layout.tsx`, maybe next skips your `loading.tsx` file because there is no layout at that level ? 🤔 i could be wrong though

**Comment by Fredkiss3** [comment]
@unleashit Normally the `Suspense` boundary is supposed to always show if your page is SSR'ed on a full page refresh, that's because of streaming (something like this : https://locaci.fredkiss.dev/search). If it don't show with SSR, then it might be a bug, it might also be because your data fetching is very fast, what throttling does is only slow down you request to the server, if it is fast enough you will sometimes get a flash of loading states, or maybe not at all since when the browser receives the response your suspense boundary has already finished loading.

You could try to add an artificial delay to your fetches of for ex 1 to 2 seconds, to see it better.

**Comment by unleashit** [comment]
> Normally the Suspense boundary is supposed to always show if your page is SSR'ed on a full page refresh

Yeah, that's exactly what I suspect the intention is but I don't think it's good UX. A better version of the Fallback would start running on client as soon as the dom is available, and/or as soon as a route transition begins if inside a layout. Then ending either as soon as content starts to paint, or when its finished (probably best). We can disagree on this, but there's no reason why it can't be implemented this way... and no reason I can think of that it wouldn't be a better user experience than a loader that appears before and/or after a pause. It already works like that in loading.tsx for the most part (seemingly with bugs), which is why you see the transition between pages on site that's static.

I get that it appears to work more nicely in situations where the server component has more work do (nice work on your site). But in other situations where the reason for the delay is network, browser parsing, etc, especially a page that already has content...  it shouldn't matter what is causing the delay when it comes to fallback behavior. With client components we can use useEffect and have that control. But right now with server components you only have the behavior the Suspense boundary offers.

Interesting what you originally pointed out about the prefetched text files on my site. Just noticed it. I think they're just the static RSC output which for some reason on a static exported site have a txt extension (normally I think they're rsc). But I could be wrong!

**Comment by Fredkiss3** [comment]
For link navigations manual `<Suspense>`, do not show until after a network request, instead the  the current page is shown (because of `useTransition`), this is because react cannot know if one of your components will suspend or not, so it makes a request to the server and as soon as it finds a suspense boundary it automatically flush out a response with the fallback, then a little while after, it shows the complete page, you might have a request that only suspend on the first render (like the page i just shared, which suspends only on first request and uses client side fetching when navigating to make it fast for navigations).

For `loading.tsx` however it seems to work a little differently. Here is what i understand :
*  next try to get the `loading.tsx` for a route in the first navigation, then for the subsequent navigations it reuses the `loading.tsx` file if it is for the same segment. 
* If the loading is on a level just upper than the page (at the same level as a parent layout), next will use that `loading.tsx` directly. 
* For a different loading state that the one in the layout, you may add another layout in a route group with your loading.tsx file
* One thing i noticed though, for loading to be instant, next needs to prefetch the page. 

I have an exemple [without prefetching](https://parallel-routes-modal-next-kvh3v93mi-fredkiss3.vercel.app/boards/oWNLDSH2nQTVrDcWS3zg7o) which shows the loading spinner only after a delay then show them instantly on subsequent navigations 

I also have an with [prefetch=true](https://parallel-routes-modal-next-git-feat-loading-re-cf3d66-fredkiss3.vercel.app/boards/oWNLDSH2nQTVrDcWS3zg7o) that shows them instantly


> it shouldn't matter what is causing the delay when it comes to fallback behavior.

I think the logic of the react team is that Suspense gives you the power to show the loading spinner whenever you like, by default it does not do exactly that, but you can take over this by using a transition, i've done something like that here : https://parallel-routes-modal-next.vercel.app/boards_search/oWNLDSH2nQTVrDcWS3zg7o

You can see here that the suspense boundary show up instantly because i suscribed to the `isPending` given by useTransition.

However this example is very hacky as i have to opt out of the link component entirely : https://github.com/Fredkiss3/parallel-routes-modal-next/blob/c9443997ebc8ccaadc03c0a68a2e8ab17ed1f260/src/app/boards_search/%5Bbid%5D/card-list.tsx#L41

I think the canonical solution here is the loading.tsx file, you just have to make sure to place them in the right place and add prefetching.

**Comment by unleashit** [comment]
I think it comes down to justifying behaviors because of what may seem to make sense "technically". But with user experience, you do your best to achieve behavior based on what the user would expect and prefer. If the cost is high, you sometimes have to compromise of course. I'm not going to have one of the first tickets to Mars for example. My UX is probably going to be going camping instead. But since I don't think this is quite rocket science and Sever Components are still new, it seems like a good time to focus on what the UX of it should be.

I think the acceptance of this is a Next thing... coming from the fact that getServersideProps was accepted as the norm even though it was always bad UX. There were always better ways to handle client route changes via lazy loading [such as this](https://github.com/ctrlplusb/react-async-component), which allowed finer control of the experience that what GSSP offered. RSCs are an upgrade over both methods, but the fallback situation is still less than ideal. It may work fine in some situations, but in others it doesn't. There's no reason why React and Next can't fix it if they want. Just show the fallback as soon as the route tranistion begins and stop showing it when the rsc is ready to hydrate. I'm not an expert on Server Components, haven't watched any two hour videos and don't currently have time, but I just can't image what the technical barrier to that could be.


So I would argue that any and all tecnically limitations

**Comment by Fredkiss3** [comment]
You are totally right about the expectations of user experiences. 

I don't know if it is "fixable" by the Next or React team, it doesn't seem like it for `Suspense` to me, but I may be totally wrong.

I'm totally in the same length as you about what we can do to make UX better. 

My suggestions (with informations I have today) : 

1 - use `Loading.tsx` for page transitions
2 - use `Suspense` for the first load if you want to have Streaming on just some components and takeover with `useTransition` to show loading fallback instantly on the client

In a live video with Ben Holmes (the whiteboard guy who works at Astro) & Dan Abramov, they made a little app with a Search list, they did the same as my second suggestion : https://github.com/bholmesdev/simple-rsc/tree/main/app

**Comment by unleashit** [comment]
I don't see why it isn't fixable. For each suspense, add a HOC to the client bundle that loads the fallback immediately on mount. Then end it once it has children. Seems like loader.js is already doing that (although not perfectly), and probably not far off from a normal suspense. I would think this is probably something for Next rather than React to do. Of course I'm being an arm chair directory who hasn't taken the time to look at the code, but if this issue is for some reason unsolvable than in my opinion is is a big mark against server components.

I appreciate the ideas! For my little project I won't sweat it too much. I've looked at useTransition and it seems more for not blocking the UI when doing some work? I'll have to check that out at some point.

**Comment by Fredkiss3** [comment]
There is a PR about a potential solution for this : https://github.com/vercel/next.js/pull/49077

TLDR : next will now prefetch pages content until it finds a `loading.tsx` which will be prefetched also. 

We just have to wait until the next canary to test and see.

**Comment by unleashit** [comment]
Thanks for the update. But this only works with prefetching? I wonder why they're against simply bundling the loaders with the initial payload. That would be a much better solution IMO. Prefetching can be expensive to a cloud hosting bill (and there is another discussion about the inability to turn it completely off on Link components), and it doesn't solve the cases where a user clicks on a link before the files are resolved. Unless I'm missing something, there would still be a delay.

**Comment by Fredkiss3** [comment]
Yes, in slow connections you can have a delay, i still think this is an interesting solution. I don't work in the next team so i can't change it any way that matters, i can understand how it is done and i have some intuition about how it all works, but i didn't dig in the source code or talked with the team for their insights maybe @feedthejim can answer ? (sorry for tagging you).

You can disable prefetching with `<Link prefetch={false}>`.

**Comment by unleashit** [comment]
Just pointing out that the way loaders were always done on SSR React apps besides Next.js, this problem didn't even exist. So now coming to Next it's new to me. Server components make it more complex, but I think the issue is really the way Next handles code splitting (more is always better). But it's not. It should really be a developer choice to optimize and decide the best tradeoff between UX and performance. Adding the loaders (maybe say down to the next route segment) in the initial bundle probably only adds a few K, and would even save a few HTTP requests. Then it works smooth as butter.

In a similar vein it's sometimes be nice to bundle a group of common pages and to download on every route. That can make the UX much more snappy as long as they aren't too expensive. I like Next, but it's kind of sad sometimes that so much is under the hood not easy to control...

> You can disable prefetching with `<Link prefetch={false}>.`

Unless they've changed it recently, you can't disable it completely. In the docs prefetch=false still triggers prefetch on hover: https://nextjs.org/docs/api-reference/next/link. If you're on a budget and paying by the request, this may not be such a good thing. Prefetching is a cool idea, but it's not always better than the selective bundling I mentioned... and certainly not more budget friendly. But enough of my opinions! My guess is this is a won't fix since they seem to really really want to maximize HTTP reqs. For some reason.

**Comment by Fredkiss3** [comment]
@unleashit please let's keep the conversation positive. I can understand that you are frustrated but don't be snarky.

> Unless they've changed it recently, you can't disable it completely.

I've tested this with the latest canary before answering. It does indeed disable prefetching.

**Comment by unleashit** [comment]
Hey, not sure what made you offended... it certainly wasn't my intention. I'm not even really frustrated. Just evaluating Next.js and server components on a simple app and made some observations. If you don't like them, that's ok,, we can disagree. That said, my opinions were only about Next.js and the decisions they've made, not you in case you felt that for some reason. Vercel is a for profit company and I respect that Next.js (and you and I) are a business model for them. I have no issues with that. But I also don't see a problem with discussing the potential pitfalls of that situation. And there are some. Just on the Link/prefetching issue for example there are numerous threads (Google and you'll find) going back several years of people not happy with the inability to turn off without creating your own <Link/>. If they fixed it recently, good on them.

Anyoo, I've 100% come in peace with the goal of discussing my experience with this issue.

**Comment by vimtor** [comment]
After several attempts, I realized that the issue for me was using `generateStaticParams`. I'm not sure why, but removing it solved the problem.

**Comment by abdelkadersettah** [comment]
> 



> After several attempts, I realized that the issue for me was using `generateStaticParams`. I'm not sure why, but removing it solved the problem.

Its worked for me to after removing the `generateStaticParams`

**Comment by rusted-love** [comment]
Nothing mentioned above helped. We expect `<Suspense>` fallback to be displayed each time page parameter changes.
E.g. `/products?cat=shoes` and  `/products?cat=shoes&size=36`. As for now server component is waiting until api requests will be completed. In case of slow search customer does not see any effect caused by link click until request is fulfilled. This is weird. 
I hope this will be fixed as soon a s possible.

**Comment by Fredkiss3** [comment]
@rusted-love i also thought this was a bug, but this is expected, the solution is to use a different key to Suspense to force React to consider the Suspense as a different component each time and show the fallback, i've done something like that here : https://github.com/Fredkiss3/todo-app-beautiful-ux/blob/main/src/app/(routes)/(app)/%40todo_app/page.tsx#L49-L54

You could either encore the search params to string to set a unique key per search, or use `Math.random()` to have a random key each time and always show the fallback. Encoding the searchParams for the key could be interesting if the user refetch the same page, since the key is the same React won't show the fallback and the user will still see the old content, while when navigating to another page, it will show them a fallback. 


I think this behavior is related to how `Suspense` work with transitions. the thing is that when using `Suspense` if the fallback have already been shown once, if you use a transition, React will hold the old data until the Suspense boundary has resolved, and Next uses transitions for all the navigations. Using a different key forces React to consider the `Suspense` as a different component.
  
This is from the React documentation here : https://react.dev/reference/react/Suspense#preventing-already-revealed-content-from-hiding

**Comment by rusted-love** [comment]
@Fredkiss3 Thank you for information. It helped me a lot.
In our case main issue was caused by slow `generateMetadata`. I thought that `generateMetadata` result is lazy-injected in order to improve TTFB metric.
Hope with time `generateMetadata` will be lazy-injected for non-robot users.

**Comment by Fredkiss3** [comment]
@rusted-love `generateMetadata` is not lazy loaded, in fact it is loaded before your page (for SEO reasons), It is explained in the docs : 

> Next.js will wait for data fetching inside generateMetadata to complete before streaming UI to the client. This guarantees the first part of a [streamed response](https://nextjs.org/docs/app/building-your-application/routing/loading-ui-and-streaming) includes <head> tags.

You can learn more here : https://nextjs.org/docs/app/building-your-application/optimizing/metadata#dynamic-metadata (in the "Good to know" section down).

**Comment by rusted-love** [comment]
@Fredkiss3 Update on my previous reply. In the development mode, the issue disappeared. However, when we deployed our website to production, it turned out that nothing had changed.
It's weird that the app directory currently exhibits such different behavior between production and development.

> @rusted-love i also thought this was a bug, but this is expected, the solution is to use a different key to Suspense to force React to consider the Suspense as a different component each time and show the fallback, i've done something like that here : https://github.com/Fredkiss3/todo-app-beautiful-ux/blob/main/src/app/(routes)/(app)/%40todo_app/page.tsx#L49-L54
> 
> You could either encore the search params to string to set a unique key per search, or use `Math.random()` to have a random key each time and always show the fallback. Encoding the searchParams for the key could be interesting if the user refetch the same page, since the key is the same React won't show the fallback and the user will still see the old content, while when navigating to another page, it will show them a fallback.
> 
> I think this behavior is related to how `Suspense` work with transitions. the thing is that when using `Suspense` if the fallback have already been shown once, if you use a transition, React will hold the old data until the Suspense boundary has resolved, and Next uses transitions for all the navigations. Using a different key forces React to consider the `Suspense` as a different component.
> 
> This is from the React documentation here : https://react.dev/reference/react/Suspense#preventing-already-revealed-content-from-hiding

**Comment by Fredkiss3** [comment]
Something like this happened to me, but it was a false alarm : i had to reload the browser at least 3 or 4 times to see the `Suspense` fallback. i used vercel so my assumption is that the cache had not updated yet.

You have to try to build locally and run `next start` and see the result in your local machine, if it works there then the problem might be on your host (maybe the cache or something).


It should work both on development and production. the code sample i gave you is deployed on vercel and it displays fallbacks on client navigation.

**Comment by rusted-love** [comment]
@Fredkiss3 This is bug in NextJs.
It persists on the local production server.
Additionally, we have debugged the dev server multiple times, and sometimes it exhibits different behavior with <Suspense/> for the same pages.

**Comment by Fredkiss3** [comment]
Sorry i may have lost the context with all these comments, the issue you have is that `generateMetada` is slow or is it the one where the `Suspense` fallback do not show between navigations ?

**Comment by rusted-love** [comment]
@Fredkiss3 the issue with suspense. It already discussed before by other participants.
We decided to revert our app to the client side rendering for all routes.

**Comment by unleashit** [comment]
Just wondering if there is anyone that disagrees that the "ideal" behavior for a suspense boundary (either loading.js or manual) would be to show a fallback whenever its children aren't available, regardless of the reason being work is being done on the server, network delay, etc. If you have a reasoning this wouldn't be the "ideal" (technical issues aside),  it would be interesting to hear.

In my mind anything short of fallbacks firing as soon as the parent DOM is available (during initial load or client transition) up until the at least the time the children begin to render on the screen is a bug. Sometimes a serious show stopping bug depending on the case. Updates based on search filters, etc. should probably be the same, although I feel at least in this case you have options to manually control the behavior if you like.

**Comment by StampixSMO** [comment]
Two cents from our practical implementation side: we are noticing significant delays after clicking a button that performs `router.push()`. In the old `pages` setup, you would usually set up a global router listener to show a loading state, but there's no recommended built-in way to do this with the `app` directory.

From a UX perspective, you just clicked a button and nothing happens for a while until suddenly your page jumps to a new one 2-3s later.

**Comment by DoctorManhattan123** [comment]
Hey I have the same issue, is there maybe any way to work around (so the moment I click on the link, some kind of loading state is shown). I wrote (almost) everything with server side components, so my app works properly without (almost) any JavaScript enabled in the browser. Is there any (maybe even dirty) workaround, so I do not have to migrate back to pages directory?

**Comment by Fredkiss3** [comment]
@StampixSMO If you are manually call `router.push` you can wrap your call with `useTransition`, it will give you a `isPending` state that you can use to show a loading state.

An example usage : 
```tsx
import { useTransition } from "react" 

const Component = () => {
  const [isPending, startTransition] = useTransition()

  return <button onClick={() => startTransition(() => router.push("/destination")))}>{isPending ? "Navigating...": "Go to /destination"}</button>
}
```

**Comment by DoctorManhattan123** [comment]
So my current component inside app/page.tsx looks like this:
```
export default function RedirectLoginCard() {
  return (
    <Link href={"/login"} className=" group justify-self-center hover:brightness-95">
      ...
    </Link>
  );
}
```

And I have a `loading.tsx` at the same level. But when I click on this Link (if I have not clicked on it before). There is nothing happening for 1s, then the loading state of the `loading.tsx` is displayed. I tried it out on the latest next stable release and on canary. It is the same on both versions. I can try to convert all of the things using `Link` to client side and try the workaround with `useTransition` mentioned above.

**Comment by Fredkiss3** [comment]
> @Fredkiss3 the issue with suspense. It already discussed before by other participants.
> 
> We decided to revert our app to the client side rendering for all routes.
> 

@rusted-love i am sorry for that, there are also another bug with prefetching (this : https://github.com/vercel/next.js/issues/42991) where if next have already prefetched the page, it won't show any loading state or `Suspense` fallback after subsequent navigations, even for dynamic pages, with the semantics being that to recall the server, you have to wait for 30 seconds, and the timer is reinitialised if you navigate to the same page before the 30 secs have finished.

It may be because of that, though I don't know your case and this issue happens only when you use a `Link` component. If this is not your case, I would want to see if you can give a little reproduction as I want to test.

Thank you.

**Comment by Fredkiss3** [comment]
> And I have a loading.tsx at the same level. But when I click on this Link (if I have not clicked on it before). There is nothing happening for 1s, then the loading state of the loading.tsx is displayed. I tried it out on the latest next stable release and on canary. It is the same on both versions. I can try to convert all of the things using Link to client side and try the workaround with useTransition mentioned above.

@DoctorManhattan123 is the delay also happening in production ? also, is the delay faster in local (production) or just on your host ? 

If this is the case, is there any task happening that takes `1s` on your `/login` page (like a slow fetch or something)  ?

**Comment by DoctorManhattan123** [comment]
@Fredkiss3 So I after I read the [Link docs](https://nextjs.org/docs/pages/api-reference/components/link#prefetch) again. And after I build it in production the delay was not happening anymore. But I have to look at it again when we deploy on vercel, but it should work now.

**Comment by janprasil** [comment]
I challenged the same problem. Thanks to _useTransition_ I was able to add a loader before redirect to requested page. As you can see on the attached video with simulated fast 3G connection - when loading.tsx is being fetched it shows a loader instead of the link, then it shows loader from loading.tsx until it's fully loaded.

https://github.com/vercel/next.js/assets/8022830/d9044b48-ce74-4e4e-9d00-9db123ee0b4f

**Comment by mhesham32** [comment]
> @rusted-love i also thought this was a bug, but this is expected, the solution is to use a different key to Suspense to force React to consider the Suspense as a different component each time and show the fallback, i've done something like that here : https://github.com/Fredkiss3/todo-app-beautiful-ux/blob/main/src/app/(routes)/(app)/%40todo_app/page.tsx#L49-L54
> 
> You could either encore the search params to string to set a unique key per search, or use `Math.random()` to have a random key each time and always show the fallback. Encoding the searchParams for the key could be interesting if the user refetch the same page, since the key is the same React won't show the fallback and the user will still see the old content, while when navigating to another page, it will show them a fallback.
> 
> I think this behavior is related to how `Suspense` work with transitions. the thing is that when using `Suspense` if the fallback have already been shown once, if you use a transition, React will hold the old data until the Suspense boundary has resolved, and Next uses transitions for all the navigations. Using a different key forces React to consider the `Suspense` as a different component.
> 
> This is from the React documentation here : https://react.dev/reference/react/Suspense#preventing-already-revealed-content-from-hiding

hey, thank you for putting this here but when I tried to do the same it didn't work at all the page is frozen till the request is done and then it shows the new results right after that

```ts
async function SearchPage(props: Props) {
  ...code 
  
  if (products === null) return notFound();
  const keyString = `searchMode=${props.params.searchMode}&searchText=${props.searchParams.searchText}&scanId=${props.searchParams.scanId}`;

  return isSSR() ? (
    <main className={styles.search_page}>
      <Products products={products} />
    </main>
  ) : (
    <Suspense fallback={<LoadingSkeleton />} key={keyString}>
      <main className={styles.search_page}>
        <Products products={products} />
      </main>
    </Suspense>
  );
}
```
the layout of this page contains a search form that should update the search params with the user entered data on submit

**Comment by mhesham32** [comment]
right now I think it is a good solution to use useTransition so the user knows that there is something happening in the background and the app is responsive to his actions. but also I think it will be great if Next added a prop for the page.tsx/layout.tsx components that has the pending state of the useTransition so the developer can be free to show a loading state based on that 

something similar to this code from react [docs](https://react.dev/reference/react/useTransition#building-a-suspense-enabled-router)

```js
function Router() {
  const [page, setPage] = useState('/');
  const [isPending, startTransition] = useTransition();

  function navigate(url) {
    startTransition(() => {
      setPage(url);
    });
  }

  let content;
  if (page === '/') {
    content = (
      <IndexPage navigate={navigate} />
    );
  } else if (page === '/the-beatles') {
    content = (
      <ArtistPage
        artist={{
          id: 'the-beatles',
          name: 'The Beatles',
        }}
      />
    );
  }
  return (
    <Layout isPending={isPending}>
      {content}
    </Layout>
  );
}
```

**Comment by Fredkiss3** [comment]
@mhesham32 @rusted-love 

After a little bit of time, i found out the problem with `Suspense` not showing 👉🏾 The culprit is **vercel**.

I tested on different providers to validate my hypothesis : 
- on local : the `Suspense` fallback is always shown
- on vercel : the `Suspense` fallback is shown if it hadn't been shown before but after it is never shown again, even if the request is slowed down (fake `2s` delay)
- on cloudfare (with their new `@cloudfare/next-on-pages` package) : the `Suspense` fallback is always shown
- on railway :  the `Suspense` fallback is always shown


You can inspect the code here : https://github.com/Fredkiss3/parallel-routes-modal-next/blob/main/src/app/pokedex/page.tsx
link for vercel deployment : https://parallel-routes-modal-next.vercel.app/pokedex
link for cloudfare deployment : https://next-on-cf.fredkiss.dev/pokedex 
link for railway deployment : https://parallel-routes-modal-next-production.up.railway.app/pokedex

i don't have any idea as to why the fallback is not shown right away on vercel, i even tried to inspect the network tab on chrome to see if maybe the response from vercel wasn't streaming, but i did not see anything unusual, the response seemed to be streaming normally and the network tab seems to be the same between vercel, cloudfare and railway.

**Comment by mhesham32** [comment]
yeah Vercel also didn't allow me to send stream from a route.tsx file as a response it worked locally but on Vercel it didn't work so we deployed our app on cloud run and it worked there. thanks for your example it really helped a lot

**Comment by markkkkas** [comment]
Will there be any response from the Vercel team? Should we wait for a fix, or can we start migrating to another hosting provider?

**Comment by Fredkiss3** [comment]
I created an issue for the `Suspense` case : #51033

**Comment by feedthejim** [comment]
Hey, I think we identified the issue and will be rolling out a fix.

**Comment by andrewbarnesweb** [comment]
> I challenged the same problem. Thanks to _useTransition_ I was able to add a loader before redirect to requested page. As you can see on the attached video with simulated fast 3G connection - when loading.tsx is being fetched it shows a loader instead of the link, then it shows loader from loading.tsx until it's fully loaded.
> 
> 
> 
> https://github.com/vercel/next.js/assets/8022830/d9044b48-ce74-4e4e-9d00-9db123ee0b4f
> 
> 
> 
> 
> 
> 

That's a fair solution to the delay. But then to keep the good user experience of instant feedback, you need two "loading" implementations every time, which is a pain when coming from a SPA mindset.

**Comment by pabloat81** [comment]
Hi, i have a question.

I am using nextjs@latest

I hace a table and a pagination compnent (client component) with Next Links that changes the query string in xxxxx/list?page=yy

I have the loading.tsx thats is showed the first time the page loads, but i cant make it apear when i click on a pagination Link. Is the same issue here?

Also i have a search form that uses server action, the action does a router.redirect(xxxx/list?page=1&text=yyy).

Thanks.

**Comment by pabloat81** [comment]
> Suspense doesn't trigger again when `key` matches the second time, so if you push searchParams into the url you'll want to add a `key` that changes. For example `<Suspense key={`page-${pageNumber}`} fallback={<>my loading...</>}>`

Hi, i done that but isnt working:
In the page component:
 const keyString = `text=${searchParams?.text}&page=${searchParams?.page}`;
 <Suspense fallback={<Loading />} key={keyString}>
The pagination link:
<Link key={keyString} scroll={false} href={{ query: q }} >1</>

What it es working is when i make a search in a form with a server action:
<form className="flex-none gap-2" action={search}>
                    <div className="form-control">
                        <input type="search" placeholder="Nombre / email"
                            defaultValue={searchParams?.text}
                            name="text" className="input input-bordered w-24 md:w-auto" />
                    </div>
                    <button className="btn">Buscar</button></form>

 async function search(data: FormData) {
        'use server'
        redirect(`${URL_ADMIN_PAGE_DEALERS}?text=${data.get('text')}&page=1`);
    }

**Comment by huozhi** [comment]
The original reproduction is fixed when I tried with the new canary version with nextjs. The issue mentioned [here](https://github.com/vercel/next.js/issues/43548#issuecomment-1582571414) about edge runtime streaming is fixed on vercel platform now.

Let me know if you still have any further problem, or if there's new issue encountered, please file a new issue with reproduction. Thanks y'all for the information and investigation.

**Comment by adarsh-drishya** [comment]
i am still facing this issue, my page is getting loaded after five clicks, using app router Link from 'next/navigation' pls do something

**Comment by gudvinga** [comment]
> i am still facing this issue, my page is getting loaded after five clicks, using app router Link from 'next/navigation' pls do something, suggest me fix.

Hi, just like an idea

We can navigate in two ways: router and next/link
Create some state with a loading flag, when clicking a link or navigating a router, set it to true, then when the pathname and searchParams change set it to false and use it to display the loading state in the UI.

useAppRouter.ts
```
import { useRouter as useNextRouter } from "next/navigation";

export function useAppRouter() {
  const router = useNextRouter();

  return Object.keys(router).reduce((acc, key) => {
    acc[key] = (...args) => {
      setIsLoading(true)
      router[key](...args);
    };
    return acc;
  }, {} as typeof router);
}

```
appLink.tsx
```
"use client"
import NextLink from "next/link";

export function AppLink(props) {
  return (
    <NextLink {...props} onClick={() => setIsLoading(true)}>
      {props.children}
    </NextLink>
  );
}

```

[Next.js router event example](https://nextjs.org/docs/app/api-reference/functions/use-router#router-events)
```
'use client'
 
import { useEffect } from 'react'
import { usePathname, useSearchParams } from 'next/navigation'
 
export function NavigationEvents() {
  const pathname = usePathname()
  const searchParams = useSearchParams()
 
  useEffect(() => {
    setIsLoading(false)
  }, [pathname, searchParams])
 
  return null
}
```

A terrible decision, but so far nothing better comes to mind

**Comment by leandronorcio** [comment]
The `loading.tsx` still does not appear instantly on the latest canary.


https://github.com/vercel/next.js/assets/32889996/96192e76-2554-426b-8fdb-110d0bc78a2b

**Comment by leandronorcio** [comment]
It's still hitting the server before showing the `loading.tsx` on the latest canary.

https://github.com/vercel/next.js/assets/32889996/487b9973-74f0-4dfa-9211-0296e4182ae9

**Comment by louisthomaspro** [comment]
I noticed that after the [30s prefetch cache for dynamic routes](https://nextjs.org/docs/app/building-your-application/routing/linking-and-navigating#1-prefetching), the loading.tsx is not showing up.

**Comment by arackaf** [comment]
@huozhi @timneutkens 

Can we re-open this issue? I'm on latest Next and this does definitely still appear to be an issue. The "instant" loading pages appear to clearly require a network roundtrip to display, and so seem to be anything but "instant"

Here's a demo

![2023-09-23 16 44 05](https://github.com/vercel/next.js/assets/11261266/76bb6a3f-d442-461c-845a-accead9207b5)

The related repo is here: https://github.com/arackaf/next-instant-loading 

which is deployed here: https://next-instant-loading.vercel.app/stream

**Comment by feedthejim** [comment]
Hello @arackaf thanks for reporting the issue, I attached a PR and fixed the issue. Tested here  https://next-instant-loading-8nv6psqod-feedthejim.vercel.app

**Comment by timneutkens** [comment]
Landed in https://github.com/vercel/next.js/pull/55950

**Comment by arackaf** [comment]
@timneutkens Was this released with the recent 13.5.4 version that dropped 3 days ago?

Repo / branch are (instant loading branch) https://github.com/arackaf/next-instant-loading/tree/feature/instant-loading 
![2023-10-05 21 21 55](https://github.com/vercel/next.js/assets/11261266/fd971084-aa04-4228-a849-eb2ef2b9d71b)

**Comment by jvandenaardweg** [comment]
It landed in 13.5.3: https://github.com/vercel/next.js/releases/tag/v13.5.3 , see last item in "Core changes"

But I'm also experiencing the same as @arackaf shows

**Comment by feedthejim** [comment]
So, there was a slight mix-up here: I misread @arackaf original message and I hadn't realised what I fixed was slightly different from what @arackaf was asking about. 

To clarify: to show the loading state, Next.js will always need to make a pre-emptive prefetch request to the page that you want to navigate to. In the scenario in which you haven't made that request yet, yes, Next.js won't show a loading state until that prefetch resolves. However, the payload that contains the loading state is most often static so this would only be a round trip to the CDN, 32ms for me locally, and so you wouldn't see this happen in the wild often I think, unless the user clicks on a link really quickly.

As the loading states are defined at the layout level, the sole alternative to that would be for Next.js to bundle all loading states into every page or for Next.js to be able to scan every links on a page to pre-emptively attach the loading states for it.

**Comment by jvandenaardweg** [comment]
It depends on your connection. Slow connection makes the whole app feel sluggish and not instant on route changes.

I'm currently in the process of switching over to the App directory, coming from the pages directory. And altough I see the benefits the new App dir offers, this issue is what really makes me consider moving everything back. Because the overall UX just feels slower 😞

The slowness comes down to this extra fetch that needs to resolve before routing to a different page.

> As the loading states are defined at the layout level, the sole alternative to that would be for Next.js to bundle all loading states into every page or for Next.js to be able to scan every links on a page to pre-emptively attach the loading states for it.
I just don't know enough about the NextJS and React internals of this. But it seems like the only logical fix to have that in the client bundle in order for it to be instant.

**Comment by murilo1008** [comment]
@jvandenaardweg Pleasee!

I really need to solve this problem in my project. Iam using app folder and the navigation isnt immediately.

If someone solved this problem please tell us.

**Comment by rusted-love** [comment]
@murilo1008 We decided to not use RSC at all. Also, we use NextJS 13.4.19 for a 2 months. This version has best navigation speed for us.

**Comment by arackaf** [comment]
I’m still a fan of app directory, but honestly the “instant” loading feature feels misnamed, and I’m more or less convinced it shouldn’t be used at all if it’s not actually instant.

CDN request or not it cannot ever be considered “instant” if a network request is required in order to load and show it.

I’d love to just see these loading states be automatically included in all client bundles, with devs being responsible for understanding that bundles can grow if they build too many of them, with too much content.

**Comment by arackaf** [comment]
Can I suggest Vercel add an ability to do something like 

```
export const include = true
```

to a loading file to force it to be included in client bundles, to enable them to be actually instant?

**Comment by unleashit** [comment]
> But it seems like the only logical fix is to have that in the client bundle in order for it to be instant.

Yes and 

> but honestly the “instant” loading feature feels misnamed, and I’m more or less convinced it shouldn’t be used at all if it’s not actually instant.

Both of these are the bottom line, and what I (and some others) tried to argue months ago. 

> export const include = true

This would be the opt-out method, which be a big improvement since it would fix more than this issue and give the dev a choice on how they want to balance performance (faster initial load vs faster transitions without relying on prefetch). Personally I prefer the (traditional) opt-in approach to code splitting, which is really as simple as a dynamic import per component (or page) you want to code split/lazy load.

**Comment by jvandenaardweg** [comment]
Why would you want this to be configurable? Loading states should be instant, not after a fetch to whatever server. Its just bad UX.

**Comment by feedthejim** [comment]
I wanted to clarify a bit: the fetch happens pre-emptively when the page loads, not when you navigate.

Can anybody send me a repro where their navigation is slow? I would love to investigate

**Comment by unleashit** [comment]
@jvandenaardweg Configurable in some way (either as opt-in or opt-out of bundling) I think is important. If you automatically bundle every possible fallback/loading.js in an initial load, it could get bloated in some cases. But in my comment above, I was thinking about the bigger picture of having more control over bundling in general.

> I wanted to clarify a bit: the fetch happens pre-emptively when the page loads

I think you're referring to preloading. Preloading is unreliable because it can either be turned off or the user can click a link before the response(s) resolve.

**Comment by jvandenaardweg** [comment]
> I wanted to clarify a bit: the fetch happens pre-emptively when the page loads, not when you navigate.
> 
> 
> 
> Can anybody send me a repro where their navigation is slow? I would love to investigate. Any datapoints would help us evaluate or not if we should change that part of the architecture.

It happens when you navigate, so clicking a link to a different page where that page is expected to show a loading indicator. The above gif of @arackaf  shows exactly what happens. He also has a reproduction repo.

**Comment by IGassmann** [comment]
> @timneutkens Was this released with the recent 13.5.4 version that dropped 3 days ago? I'm still seeing the same behavior.
> 
> Repo / branch are here [arackaf/next-instant-loading@`feature`/instant-loading](https://github.com/arackaf/next-instant-loading/tree/feature/instant-loading?rgh-link-date=2023-10-06T02%3A22%3A51Z)
> 
> ![2023-10-05 21 21 55](https://user-images.githubusercontent.com/11261266/273087348-fd971084-aa04-4228-a849-eb2ef2b9d71b.gif) [ ![2023-10-05 21 21 55](https://user-images.githubusercontent.com/11261266/273087348-fd971084-aa04-4228-a849-eb2ef2b9d71b.gif) ](https://user-images.githubusercontent.com/11261266/273087348-fd971084-aa04-4228-a849-eb2ef2b9d71b.gif) [ ](https://user-images.githubusercontent.com/11261266/273087348-fd971084-aa04-4228-a849-eb2ef2b9d71b.gif)

@arackaf In your reproduction, it seems that you're running the app in development mode (`next dev`), and not in production mode (`next build; next start`). As [per the docs](https://nextjs.org/docs/app/building-your-application/routing/linking-and-navigating#1-prefetching), prefetching is disabled in development. If you run your app in production mode, the loading is indeed instant after the prefetch has completed:

https://github.com/vercel/next.js/assets/4291707/de68db97-2d08-4b15-ac13-a3dc594dea15

In production, routes are automatically prefetched as they become visible in the user's viewport. Prefetching happens when the page first loads or when it comes into view through scrolling.

**Comment by IGassmann** [comment]
The fact that prefetching is not enabled in development is a common source of confusion on how this works. @feedthejim why is it disabled in development?

**Comment by unleashit** [comment]
So I guess people in the Next.js community feel that prefetching is a good solution. But a) it can be turned off in a Link and b) the user can act faster than the response comes in. This is in production. Lastly, it costs $ especially when serverless.

**Comment by feedthejim** [comment]
@unleashit I'm asking for a real repro where I don't have to simulate 3G conditions to notice a delay

@IGassmann we disable prefetching in dev because it would force the compilation of all the pages that you're linking to from your page. Please do note that if you do click a link in dev, we simulate the prefetching behaviour however by fetching the loading state first and then the rest.

**Comment by IGassmann** [comment]
> @unleashit I'm asking for a real repro where I don't have to simulate 3G conditions to notice a delay

@feedthejim, this is a challenging request. Users may encounter poor internet connections, such as when using the app in the subway. For example, a user might load a page right before entering a tunnel and then, while in the tunnel, click a navigation link, which could result in no loading indicator. How do you provide a reproduction for this situation?

**Comment by unleashit** [comment]
@feedthejim sorry, I don't have time at the moment. I do have a non-minimal statically exported repo I could show you that never shows the loaders: https://github.com/unleashit/portfolio-v4/tree/master/app/portfolio. They were actually working at some point during the early app directory development, but the behavior changed several times and now nothing.

I think most people here are talking about SSR, where the loader can sometimes be delayed like in @arackaf s demo. Again, I used to get this behavior during the 13 beta for my static site, but now nothing. But I've also seen it in SSR sites.... just don't have any at hand to post. You'll need to throttle it though to see, unless you want to take your laptop maybe a block from the house/office until the connection is bad ;)

**Comment by jvandenaardweg** [comment]
The issue also happens in production environments. The same RSC call delays rendering of a loading indicator.

> In production, routes are automatically prefetched as they become visible in the user's viewport. Prefetching happens when the page first loads or when it comes into view through scrolling.

I see the prefetching happen for the pages that are linked to. But the same fetch happens when a user clicks the link to that page.

1. Hard refresh the current page you are in
2. You'll notice the prefetching happening.
3. On the current page, click a link to a different page, let's say it links to /page/whatever
4. When clicked, a fetch happens in the background to /page/whatever?_rsc=17rda. It's the same fetch that already happened on page load.

Now, nothing happens on the screen until that fetch resolves. Not even a loading indicator. Nothing.

6. When it resolved, the loading indicator of the page is shown
7. After the RSC is done fetching some external/database data, the component is show

**Comment by jvandenaardweg** [comment]
I checked @arackaf 's repo and environment. This page shows the same issue i'm also seeing: https://next-instant-loading.vercel.app/

**Comment by IGassmann** [comment]
@jvandenaardweg something is weird; I'm able to reproduce the issue on the app deployed on Vercel ([next-instant-loading.vercel.app](https://next-instant-loading.vercel.app/)), but not when it's running locally in production mode ([arackaf/next-instant-loading@feature/instant-loading](https://github.com/arackaf/next-instant-loading/tree/feature/instant-loading?rgh-link-date=2023-10-10T10%3A02%3A09Z))

**On Vercel**
https://github.com/vercel/next.js/assets/4291707/1fcc9092-352e-4db1-aaed-9f64e1713cff

**Locally**
https://github.com/vercel/next.js/assets/4291707/68c7f5bd-12ca-4bbb-a67a-d9cd1f2ed9ba

**Comment by IGassmann** [comment]
@jvandenaardweg that successful deployment in the Preview environment is hosted at https://next-instant-loading-o7zx8tsok-adam-rackis.vercel.app/, not at https://next-instant-loading.vercel.app/

On that deployment, we can see the issue fixed:

https://github.com/vercel/next.js/assets/4291707/5c75599b-b9c0-49bc-8a58-9eeb409e46c7

**Comment by jvandenaardweg** [comment]
Thanks for clearing that up @IGassmann ! I'll then have to take another look at my code then. Because i'm on the same latest Next version but dó see the delay. I'm just totally lost now what's causing it.

**Comment by jvandenaardweg** [comment]
I guess my issue lies in the use of <Suspense> on a page to have fine grained control over what component is loading, instead of a page wide loading indicator.

Using Vercel's app router demo to demonstrate: 

No issue here when using a generic loading.tsx file: https://app-router.vercel.app/loading

But the issue is here "Streaming with Suspense": https://app-router.vercel.app/streaming

When you throttle the connection, using loading.tsx instantly shows a indicator. Which is good. But using the same throttled connection on the Streaming With Suspense prevents anything from being shown.

Both also tested with the latest NextJS version locally and with a deploy on Vercel.

So the issue lies in the use of <Suspense> with a fallback on a page.tsx. When I switch over to using the loading.tsx the issue dissapears.

**Comment by arackaf** [comment]
Ok I've updated my original repro with the latest Next version, and the instant loading state does show instantly, given the prefetch. 

I do still think loading into client bundles would be better, but it does look like this is essentially fixed and working as designed. 

https://next-instant-loading.vercel.app/no-suspense

**Comment by unleashit** [comment]
It may be working as designed, but it doesn't work. Relying on a lazy loaded response can and will lead to weird glitches with loaders. Separately, I would expect loading.js (and IMO suspense fallbacks) even in static/SSG sites to show the fallback if no RSC for the page or component has yet been downloaded, which can happen when prefetch prop is set to false.

It's hard to not get the feeling that Vercel has planned for a heavy use of prefetching for their own concerns and we has devs should not just brush this under the rug and give them a pass. We are more or less forced by the industry to learn and use this tool, but it can be made better if you aren't afraid to give them a little pressure about it. Prefetching all the things is a terrible idea. Worrying about adding 10-20k to a bundle is good, but never without consideration about the overall balance.

**Comment by murilo1008** [comment]
I updated my project to the latest version of Next. In fact, loading has improved along with page change speed, but it is not yet in immediate development.

I used this command: 

npm i next@latest react@latest react-dom@latest eslint-config-next@latest


Do I need to do anything else to make page switching instantaneous?

**Comment by arackaf** [comment]
@murilo1008 I'd create a simple repro and upload to Vercel. Feel free to fork mine above

**Comment by leerob** [comment]
I want to clarify how loading UI works in Next.js.

Routing in the App Router is "server centric". We need a network roundtrip to the server when doing route transitions. The client doesn't know what loading state to render, until it communicates with the server.

When we prefetch, we can retrieve both the route to navigate to _and_ the loading UI. Prefetching is what makes navigations feel snappy. If you prefetch the loading UI which does not have data fetching, then you can "instantly" show this loading state, followed by the dynamic content on the page being streamed in after. This is enabled through React Suspense and streaming server-rendering as part of the App Router.

We don't believe an alternate solution where you include all of the routing variants on the client is feasible or the best solution for Next.js. For example, consider a loading state which is `100kb` of HTML / possible routes. Should that be included in the initial JavaScript bundle, blocking interactivity for *everything* on the page until it's downloaded? Or, is it better to *not* block interactivity, and load it asynchronously before the next click? We still believe the current model is correct.

There's still opportunities for us to improve the experience of transitioning between loading states. For example, we can optimize further the de-duplication of data. Thanks for the reproductions here, as they'll help us dig in further and find places to optimize.

**Comment by Systemcluster** [comment]
> Routing in the App Router is "server centric". We need a network roundtrip to the server when doing route transitions. The client doesn't know what loading state to render, until it communicates with the server.

Does this include loading states from already loaded layouts? I have a single `loading.tsx` with `"use client"` inside a `layout.tsx` with `"use client"` that is used across the whole app. Every route uses this loader. I would expect the already-loaded globally-used loader to be shown immediately without prior network roundtrip.

**Comment by unleashit** [comment]
Thanks for explaining the thinking. I also just came across [Deep Dive: Caching and Revalidating]( https://github.com/vercel/next.js/discussions/54075) which is also very helpful and appreciated, The docs (at least at the point I was reading them several months ago) don't go far enough into some of these subjects or explain the thinking behind them.

Personally, I wish Next.js kept to its original roots of Isomorphic code with getInitialProps. getServersideProps started this whole thing of letting Next mange client side cache which usually results in excess server requests. That I think defeats the main advantages of SPAs, which is offloading some of the expensive work to the client and the clean separation of backend concerns. I love most of the other things that got added (SSG, even the app directory and RSCs "in the right situations"), but getInitialProps was actively discouraged which IMO is mostly why the tide started turning back to the server. Not just with Next.Js, but that's when Remix, etc. started coming out. Anyhoo, it appears the same foundation is why we get loader glitches sometimes.
